runtime/onert/core/src/util/shapeinf/While.cc

```diff
@@ -23,18 +23,120 @@ namespace shape_inference
 
 void StaticInferer::visit(const ir::operation::While &op)
 {
-  for (const auto input_idx : op.getInputs())
+  auto &cond_graph = _lowered_subgs.at(op.param().cond_subg_index)->graph();
+  auto &body_graph = _lowered_subgs.at(op.param().body_subg_index)->graph();
+  const auto inputs = op.getInputs();
+  const auto &outputs = op.getOutputs();
+
+  // re-sizing input shapes of then subgraph
+  const auto &cond_inputs = cond_graph.getInputs();
+  assert(inputs.size() == cond_inputs.size());
+  for (size_t i = 0; i < inputs.size(); ++i)
   {
-    if (_operands.at(input_idx).info().isDynamic())
+    const auto &input = _operands.at(inputs.at(i));
+    auto &cond_input = cond_graph.operands().at(cond_inputs.at(i));
+    if (input.info().isDynamic())
+    {
+      cond_input.info().setDynamic();
+    }
+    else
+    {
+      auto new_shape = input.info().shape();
+      cond_input.info().shape(new_shape);
+    }
+  }
+
+  // re-sizing input shapes of body subgraph
+  const auto &body_inputs = body_graph.getInputs();
+  assert(cond_inputs.size() == body_inputs.size());
+  for (size_t i = 0; i < cond_inputs.size(); ++i)
+  {
+    const auto &cond_input = cond_graph.operands().at(cond_inputs.at(i));
+    auto &body_input = body_graph.operands().at(body_inputs.at(i));
+    if (cond_input.info().isDynamic())
+    {
+      body_input.info().setDynamic();
+    }
+    else
+    {
+      const auto &new_shape = cond_input.info().shape();
+      body_input.info().shape(new_shape);
+    }
+  }
+
+  // re-sizing operands of body subgraph
+  StaticInferer body_inferer(op.param().body_subg_index, _lowered_subgs);
+  _lowered_subgs.at(op.param().body_subg_index)
+      ->iterateTopolOpSeqs([&](const ir::OpSequenceIndex &, const ir::OpSequence &op_seq) {
+        body_inferer.infer(op_seq);
+      });
+
+  // Check whether while operation's shapes are predictable
+  // If any of shape of body outputs and cond inputs are different, non-constant operands would be
+  // set to dynamic
+  bool check_unpredictable_dynamic = false;
+  const auto &body_outputs = body_graph.getOutputs();
+  assert(body_outputs.size() == cond_inputs.size());
+  for (size_t i = 0; i < body_outputs.size(); ++i)
+  {
+    const auto &body_output = body_graph.operands().at(body_outputs.at(i));
+    auto &cond_input = cond_graph.operands().at(cond_inputs.at(i));
+    if ((cond_input.info().isDynamic() != body_output.info().isDynamic()) ||
+        (cond_input.shape() != body_output.shape()))
+    {
+      check_unpredictable_dynamic = true;
+      break;
+    }
+  }
+
+  if (check_unpredictable_dynamic)
+  {
+    // Set inputs of body subgraph
```

Comment:
when only `i`th input of cond and while was _unpredictable_,
the code below set _all_ inputs dynamic.
Could you explain the reason?

Reply:
While operation makes body subgraph looping. We cannot predict how many the subgraph loops and which outputs change to dynamic by the loop. So I added the code to set all inputs to dynamic if one of outputs change to dynamic.